File: Python/Strings/2292 - Painel LED.py
"""
Autor: Pierre Vieira
Data da submissão: 02/02/2020 15:55:51
"""


def decidir(elemento):
    if elemento[0] == 'X' and elemento[1] % 2 == 0:
        return 'X'
    elif elemento[0] == 'X' and elemento[1] % 2 == 1:
        return 'O'
    elif elemento[0] == 'O' and elemento[1] % 2 == 0:
        return 'O'
    return 'X'


def solve_this_problem(string, qtde_piscadas):
    lista_qtde_vezes_trocou_de_estado = [qtde_piscadas]
    lista_qtde_vezes_apagou = [qtde_piscadas // 2 if string[0] == 'X' else (qtde_piscadas + 1) // 2]
    for c in range(1, len(string)):
        lista_qtde_vezes_trocou_de_estado.append(lista_qtde_vezes_apagou[c - 1])
        lista_qtde_vezes_apagou.append(lista_qtde_vezes_trocou_de_estado[-1] // 2 if string[c] == 'X' else (lista_qtde_vezes_trocou_de_estado[-1] + 1) // 2)
    correspondencia = tuple(zip(list(string), lista_qtde_vezes_trocou_de_estado))
    return ''.join(list(map(decidir, correspondencia)))


def main():
    for c in range(int(input())):
        linha = input().split()
        inicio, qtde_piscadas = linha[0], int(linha[1])
        print(solve_this_problem(inicio, qtde_piscadas))


main()

File: Kotlin/Strings/2292 - Painel LED.kt
/*
* Autor: Pierre Vieira
* Data de submissão: 20/03/2023 02:22:48
*/

fun main(args: Array<String>) {
    repeat(readLine()!!.toInt()) {
        val line = readLine()!!.split(" ")
        val initialConfig = line.first()
        val c = line[1].toLong()
        println(solve(initialConfig, c))
    }
}

fun decide(element: Pair<Char, Long>): Char {
    return when {
        element.first == 'X' && element.second % 2 == 0L -> 'X'
        element.first == 'X' && element.second % 2 == 1L -> 'O'
        element.first == 'O' && element.second % 2 == 0L -> 'O'
        else -> 'X'
    }
}

fun solve(initialConfig: String, blinks: Long): String {
    val blinkCountList = mutableListOf(blinks)
    val offCountList = mutableListOf(if (initialConfig[0] == 'X') blinks / 2 else (blinks + 1) / 2)

    for (i in 1 until initialConfig.length) {
        blinkCountList.add(offCountList[i - 1])
        val lastBlink = blinkCountList.last()
        offCountList.add(if (initialConfig[i] == 'X') lastBlink / 2 else (lastBlink + 1) / 2)
    }

    val mapping = initialConfig.mapIndexed { index, char -> char to blinkCountList[index] }
    return mapping.map(::decide).joinToString("")
}